File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/widgets/FirstRunView.java
package com.battlelancer.seriesguide.widgets;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.annotation.IntDef;
import android.support.v7.widget.CardView;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.ImageButton;
import android.widget.RelativeLayout;
import butterknife.ButterKnife;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.settings.DisplaySettings;
import com.battlelancer.seriesguide.util.TaskManager;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import org.greenrobot.eventbus.EventBus;

public class FirstRunView extends CardView {

    private static final String PREF_KEY_FIRSTRUN = "accepted_eula";

    @IntDef({ ButtonType.DISMISS,
            ButtonType.ADD_SHOW,
            ButtonType.SIGN_IN,
            ButtonType.RESTORE_BACKUP })
    @Retention(RetentionPolicy.SOURCE)
    public @interface ButtonType {
        int DISMISS = 0;
        int ADD_SHOW = 1;
        int SIGN_IN = 2;
        int RESTORE_BACKUP = 3;
    }

    public static class ButtonEvent {
        public final FirstRunView firstRunView;
        @ButtonType public final int type;

        public ButtonEvent(FirstRunView firstRunView, @ButtonType int type) {
            this.firstRunView = firstRunView;
            this.type = type;
        }
    }

    public static boolean hasSeenFirstRunFragment(final Context context) {
        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(context);
        return sp.getBoolean(PREF_KEY_FIRSTRUN, false);
    }

    public FirstRunView(Context context) {
        this(context, null);
    }

    public FirstRunView(Context context, AttributeSet attrs) {
        super(context, attrs);

        LayoutInflater.from(context).inflate(R.layout.view_first_run, this);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        RelativeLayout noSpoilerView = ButterKnife.findById(this, R.id.containerFirstRunNoSpoilers);
        final CheckBox noSpoilerCheckBox = ButterKnife.findById(noSpoilerView,
                R.id.checkboxFirstRunNoSpoilers);
        Button buttonAddShow = ButterKnife.findById(this, R.id.buttonFirstRunAddShow);
        Button buttonSignIn = ButterKnife.findById(this, R.id.buttonFirstRunSignIn);
        Button buttonRestoreBackup = ButterKnife.findById(this, R.id.buttonFirstRunRestore);
        ImageButton buttonDismiss = ButterKnife.findById(this, R.id.buttonFirstRunDismiss);

        noSpoilerView.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                // new state is inversion of current state
                boolean noSpoilers = !noSpoilerCheckBox.isChecked();
                // save
                PreferenceManager.getDefaultSharedPreferences(v.getContext()).edit()
                        .putBoolean(DisplaySettings.KEY_PREVENT_SPOILERS, noSpoilers)
                        .apply();
                // update next episode strings right away
                TaskManager.getInstance(v.getContext()).tryNextEpisodeUpdateTask();
                // show
                noSpoilerCheckBox.setChecked(noSpoilers);
            }
        });
        noSpoilerCheckBox.setChecked(DisplaySettings.preventSpoilers(getContext()));
        buttonAddShow.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                EventBus.getDefault().post(new ButtonEvent(FirstRunView.this, ButtonType.ADD_SHOW));
            }
        });
        buttonSignIn.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                EventBus.getDefault()
                        .post(new ButtonEvent(FirstRunView.this, ButtonType.SIGN_IN));
            }
        });
        buttonRestoreBackup.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                EventBus.getDefault()
                        .post(new ButtonEvent(FirstRunView.this, ButtonType.RESTORE_BACKUP));
            }
        });
        buttonDismiss.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                setFirstRunDismissed();
                EventBus.getDefault().post(new ButtonEvent(FirstRunView.this, ButtonType.DISMISS));
            }
        });
    }

    private void setFirstRunDismissed() {
        SharedPreferences prefs = PreferenceManager
                .getDefaultSharedPreferences(getContext());
        prefs.edit().putBoolean(FirstRunView.PREF_KEY_FIRSTRUN, true).apply();
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/shows/FirstRunView.kt
package com.battlelancer.seriesguide.shows

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.FrameLayout
import androidx.core.content.edit
import androidx.core.view.isGone
import androidx.preference.PreferenceManager
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.databinding.ViewFirstRunBinding
import com.battlelancer.seriesguide.dataliberation.AutoBackupTools
import com.battlelancer.seriesguide.settings.AppSettings
import com.battlelancer.seriesguide.settings.DisplaySettings
import com.battlelancer.seriesguide.settings.NotificationSettings
import com.battlelancer.seriesguide.settings.UpdateSettings
import com.battlelancer.seriesguide.util.TaskManager
import com.battlelancer.seriesguide.util.TextTools
import com.battlelancer.seriesguide.util.WebTools
import com.uwetrottmann.androidutils.AndroidUtils

class FirstRunView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) :
    FrameLayout(context, attrs) {

    interface FirstRunClickListener {
        fun onAddShowClicked()
        fun onSignInClicked()
        fun onRestoreBackupClicked()
        fun onRestoreAutoBackupClicked()
        fun onAllowNotificationsClicked()
        fun onAllowPreciseNotificationsClicked()
        fun onDismissClicked()
    }

    private val binding: ViewFirstRunBinding =
        ViewFirstRunBinding.inflate(LayoutInflater.from(context), this, true)

    var clickListener: FirstRunClickListener? = null

    override fun onFinishInflate() {
        super.onFinishInflate()

        binding.buttonAllowNotifications.setOnClickListener {
            clickListener?.onAllowNotificationsClicked()
        }

        binding.buttonAllowPreciseNotifications.setOnClickListener {
            clickListener?.onAllowPreciseNotificationsClicked()
        }

        binding.buttonRestoreAutoBackup.setOnClickListener {
            clickListener?.onRestoreAutoBackupClicked()
        }

        binding.containerNoSpoilers.setOnClickListener { v ->
            // new state is inversion of current state
            val noSpoilers = !binding.checkboxNoSpoilers.isChecked
            // save
            PreferenceManager.getDefaultSharedPreferences(v.context).edit {
                putBoolean(DisplaySettings.KEY_PREVENT_SPOILERS, noSpoilers)
            }
            // update next episode strings right away
            TaskManager.getInstance().tryNextEpisodeUpdateTask(v.context)
            // show
            binding.checkboxNoSpoilers.isChecked = noSpoilers
        }
        binding.checkboxNoSpoilers.text = TextTools.buildTitleAndSummary(
            context,
            R.string.pref_nospoilers,
            R.string.pref_nospoilers_summary
        )

        binding.containerDataSaver.setOnClickListener {
            val isSaveData = !binding.checkboxDataSaver.isChecked
            PreferenceManager.getDefaultSharedPreferences(it.context).edit {
                putBoolean(UpdateSettings.KEY_ONLYWIFI, isSaveData)
            }
            binding.checkboxDataSaver.isChecked = isSaveData
        }
        binding.checkboxDataSaver.text = TextTools.buildTitleAndSummary(
            context,
            R.string.pref_updatewifionly,
            R.string.pref_updatewifionlysummary
        )

        binding.buttonAddShow.setOnClickListener {
            clickListener?.onAddShowClicked()
        }
        binding.buttonSignIn.setOnClickListener {
            clickListener?.onSignInClicked()
        }
        binding.buttonRestoreBackup.setOnClickListener {
            clickListener?.onRestoreBackupClicked()
        }
        binding.buttonDismiss.setOnClickListener {
            setFirstRunDismissed()
            clickListener?.onDismissClicked()
        }

        binding.containerErrorReports.setOnClickListener {
            // New state is inversion of current state.
            val isSendErrorReports = !binding.checkboxErrorReports.isChecked
            AppSettings.setSendErrorReports(context, isSendErrorReports, true)
            binding.checkboxErrorReports.isChecked = isSendErrorReports
        }
        binding.checkboxErrorReports.text = TextTools.buildTextAppearanceSpan(
            context,
            R.string.pref_error_reports,
            R.style.TextAppearance_SeriesGuide_Subtitle1_Secondary
        )

        binding.textViewPolicyLink.setOnClickListener {
            WebTools.openInCustomTab(context, context.getString(R.string.url_privacy))
        }
    }

    fun bind() {
        binding.groupAllowNotifications.isGone =
            !AndroidUtils.isAtLeastTiramisu || NotificationSettings.areNotificationsAllowed(context)

        binding.groupAllowPreciseNotifications.isGone =
            NotificationSettings.canScheduleExactAlarms(context)

        binding.groupAutoBackupDetected.isGone =
            !AutoBackupTools.isAutoBackupMaybeAvailable(context)

        binding.checkboxNoSpoilers.isChecked = DisplaySettings.preventSpoilers(context)
        binding.checkboxDataSaver.isChecked = UpdateSettings.isLargeDataOverWifiOnly(context)
        binding.checkboxErrorReports.isChecked = AppSettings.isSendErrorReports(context)
    }

    private fun setFirstRunDismissed() {
        PreferenceManager.getDefaultSharedPreferences(context).edit {
            putBoolean(PREF_KEY_FIRSTRUN, true)
        }
    }

    companion object {

        const val PREF_KEY_FIRSTRUN = "accepted_eula"

        @JvmStatic
        fun hasSeenFirstRunFragment(context: Context): Boolean {
            val sp = PreferenceManager.getDefaultSharedPreferences(context)
            return sp.getBoolean(PREF_KEY_FIRSTRUN, false)
        }
    }
}
